tests/unit/anchore_engine/services/policy_engine/policy/gates/test_dockerfile.py

```diff
@@ -0,0 +1,278 @@
+        "expected_msg": "Dockerfile directive 'COPY' check '=' matched against '. /app' for line '. /app'",
+    },
+    {
+        "trigger_type": "instruction_check",
+        "trigger_params": {"instruction": "COPY", "check": "!=", "value": ". /app"},
+        "expected_fire": False,
+    },
+    {
+        "trigger_type": "instruction_check",
+        "trigger_params": {"instruction": "COPY", "check": "exists"},
+        "expected_fire": True,
+        "expected_msg": "Dockerfile directive 'COPY' check 'exists' matched against '' for line '. /app'",
+    },
+    {
+        "trigger_type": "instruction_check",
+        "trigger_params": {"instruction": "VOLUME", "check": "not_exists"},
+        "expected_fire": True,
+        "expected_msg": "Dockerfile directive 'VOLUME' not found, matching condition 'not_exists' check",
+    },
+    {
+        "trigger_type": "instruction_check",
+        "trigger_params": {"instruction": "COPY", "check": "not_exists"},
+        "expected_fire": False,
+    },
+    {
+        "trigger_type": "instruction_check",
+        "trigger_params": {"instruction": "COPY", "check": "like", "value": ". /"},
+        "expected_fire": True,
+        "expected_msg": "Dockerfile directive 'COPY' check 'like' matched against '. /' for line '. /app'",
+    },
+    {
+        "trigger_type": "instruction_check",
+        "trigger_params": {
+            "instruction": "COPY",
+            "check": "like",
+            "value": ". /usr/local/bin",
+        },
+        "expected_fire": False,
+    },
+    {
+        "trigger_type": "instruction_check",
+        "trigger_params": {
+            "instruction": "COPY",
+            "check": "not_like",
+            "value": ". /usr",
+        },
+        "expected_fire": True,
+        "expected_msg": "Dockerfile directive 'COPY' check 'not_like' matched against '. /usr' for line '. /app'",
+    },
+    {
+        "trigger_type": "instruction_check",
+        "trigger_params": {
+            "instruction": "COPY",
+            "check": "not_like",
+            "value": ". /app",
+        },
+        "expected_fire": False,
+    },
+]
+
+actual_dockerfile_only_tests = [
+    {
+        "trigger_type": "instruction_check",
+        "trigger_params": {
+            "actual_dockerfile_only": True,
+            "instruction": "COPY",
+            "check": "exists",
+        },
+        "expected_fire": False,
+        "dockerfile_mode": None,
+    },
+    {
+        "trigger_type": "exposed_port",
+        "trigger_params": {
+            "actual_dockerfile_only": True,
+            "ports": "3000",
+            "type": "blacklist",
+        },
+        "expected_fire": False,
+        "dockerfile_mode": None,
+    },
+]
+
+test_contexts = [
+    *no_dockerfile_trigger_tests,
+    *effective_user_trigger_tests,
+    *exposed_port_trigger_tests,
+    *instruction_check_trigger_tests,
+    *actual_dockerfile_only_tests,
+]
+
+
+@pytest.mark.parametrize("test_context", test_contexts)
```

Comment:
This is just a nit, so not blocking, but while I agree this is an efficient way to do the test, generally I'd prefer separate test functions for each trigger so that the test output is a bit easier to understand and it keeps the matrix for each test function a bit more tractable as the number of parameters and input types grows. Something to keep in mind for the future.

Overall though this looks good! Nice work!

Reply:
@zhill I already used this model for the other PR so I'll go ahead and change it for the test I am currently working on. Would the solution have been somewhat like my first commit where the test functions are more or less duplicated with the only distinction being the different trigger creators? Or would it be ok to create a function that handles the assertions that gets called from each test